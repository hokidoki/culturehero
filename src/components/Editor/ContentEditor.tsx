import React, { useCallback, useMemo, useRef, useState } from 'react';
import ContentEditable from 'react-contenteditable';
import ContentEditorStyle from './ContentEditor.styles';
import { ContentEditorProps } from './ContentEditor.types';

export default function ContentEditor({
  title: __title__ = '',
  article: __article__ = '',
  onSubmit,
}: ContentEditorProps) {
  const [title, seTtitle] = useState(__title__);
  const [article, setArticle] = useState(__article__);
  const [isLoading, setIsLoading] = useState(false);
  const ref = useRef<HTMLDivElement>(null);
  const disabled = useMemo(() => {
    const minimumLength = 3;
    const lengthCondition =
      title.length < minimumLength || article.length < minimumLength;
    const sameCondition = title === __title__ && __article__ === article;
    return sameCondition || lengthCondition;
  }, [title, article, __title__, __article__]);

  const preSubmitHandler = useCallback(
    (title: string, article: string) => {
      setIsLoading(true);
      onSubmit({ title, article }).catch(() => setIsLoading(false));
    },
    [onSubmit]
  );

  return (
    <ContentEditorStyle.Layout>
      <ContentEditorStyle.TitleInput
        placeholder={'제목을 입력하세요'}
        value={title}
        disabled={isLoading}
        onChange={(e) => seTtitle(e.target.value)}
      />
      <ContentEditorStyle.ArticleTextArea onClick={() => ref.current?.focus()}>
        <ContentEditable
          innerRef={ref}
          html={article}
          disabled={isLoading}
          onChange={(e) => setArticle(e.target.value)}
        />
      </ContentEditorStyle.ArticleTextArea>
      <ContentEditorStyle.SubmitButton
        disabled={disabled || isLoading}
        onClick={() => preSubmitHandler(title, article)}
      >
        저장
      </ContentEditorStyle.SubmitButton>
    </ContentEditorStyle.Layout>
  );
}
