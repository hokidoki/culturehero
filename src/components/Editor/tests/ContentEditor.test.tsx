import { fireEvent, render, waitFor } from '@render';
import ContentEditor from '../ContentEditor';
import { isNull } from '@fxts/core';

describe('ContentEditor', () => {
  it('match snapshot', () => {
    const props = {
      title: 'TITLE',
      article: 'ARTICLE',
      onSubmit: () => Promise.resolve(),
    };
    const utils = render(<ContentEditor {...props} />);
    expect(utils.container).toMatchSnapshot();
  });

  it('마운트 당시의 제목과 글이 이전과 동일한 경우, button은 disabled상태이다.', async () => {
    const props = {
      title: 'TITLE',
      article: 'ARTICLE',
      onSubmit: () => Promise.resolve(),
    };
    const utils = render(<ContentEditor {...props} />);

    const input = utils.container.querySelector('input');
    const editor = utils.container.querySelector('[contenteditable]');
    const button = utils.container.querySelector('button');

    if (isNull(input) || isNull(button) || isNull(editor))
      throw new Error('Element is null');

    expect(button.disabled).toBeTruthy();
    fireEvent.input(input, { target: { value: 'NEW_TITLE' } });
    expect(button.disabled).toBeFalsy();
  });

  it('입력이 3이하인 경우, button은 disabled상태이다.', async () => {
    const props = {
      title: 'TITLE',
      article: 'ARTICLE',
      onSubmit: () => Promise.resolve(),
    };
    const utils = render(<ContentEditor {...props} />);

    const input = utils.container.querySelector('input');
    const editor = utils.container.querySelector('[contenteditable]');
    const button = utils.container.querySelector('button');

    if (isNull(input) || isNull(button) || isNull(editor))
      throw new Error('Element is null');

    expect(button.disabled).toBeTruthy();
    fireEvent.input(input, { target: { value: 'N' } });
    fireEvent.input(editor, { target: { innerHTML: 'H' } });
    expect(button.disabled).toBeTruthy();
  });
  it('버튼을 눌러 입력을 제출하여 pending일 때, input과 [contenteditable]요소는 disabled 상태이다.', async () => {
    const props = {
      title: 'TITLE',
      article: 'ARTICLE',
      onSubmit: () => Promise.resolve(),
    };
    const utils = render(<ContentEditor {...props} />);

    const input = utils.container.querySelector('input');
    const editor = utils.container.querySelector('[contenteditable]');
    const button = utils.container.querySelector('button');

    if (isNull(input) || isNull(button) || isNull(editor))
      throw new Error('Element is null');

    expect(button.disabled).toBeTruthy();
    fireEvent.input(input, { target: { value: 'NEW_TITLE' } });
    fireEvent.input(editor, { target: { innerHTML: 'NEW_ARTICLE' } });
    expect(button.disabled).toBeFalsy();

    fireEvent.click(button);
    await waitFor(() => {
      expect(input.disabled).toBeTruthy();
      expect(editor.getAttribute('contenteditable')).toBe('false');
    });
  });
});
