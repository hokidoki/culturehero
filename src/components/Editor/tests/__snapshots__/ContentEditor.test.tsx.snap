// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentEditor match snapshot 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  min-height: 100%;
  background-color: rgb(17,17,20);
  padding: 40px;
  padding-top: 60px;
}

.c1 {
  width: 100%;
  color: white;
  font-size: 2.5rem;
  margin-bottom: 20px;
  height: 2.5rem;
  background-color: transparent;
  outline: none;
  border: none;
}

.c2 {
  width: 100%;
  min-height: 700px;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
  background-color: transparent;
  outline: none;
  resize: none;
  padding: 10px;
  font-size: 1rem;
  color: #9aa0a6;
  cursor: text;
}

.c2 div[contenteditable=true] {
  outline: none;
}

.c3 {
  width: 100%;
  height: 50px;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
  margin-top: 15px;
  background-color: transparent;
  font-size: 1.25rem;
  color: white;
  transition: all 0.5s;
}

.c3:disabled {
  color: #bdc1c6;
  cursor: not-allowed;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="제목을 입력하세요"
      value="TITLE"
    />
    <div
      class="c2"
    >
      <div
        contenteditable="true"
      >
        ARTICLE
      </div>
    </div>
    <button
      class="c3"
      disabled=""
    >
      저장
    </button>
  </div>
</div>
`;
